### PR TITLE
server: add a 30 minute timeout for event trigger locks

### PR DESCRIPTION
**Pull request by paf31**
This gives event trigger locks a 30 minute timeout. In case the graceful shutdown of a server fails, this will automatically unlock in-process rows after 30 minutes at most. 

This PR is a work in progress and I haven't had chance to test it extensively yet, but we can use this as a place to discuss the approach.

### Changelog

- [ ] `CHANGELOG.md` is updated with user-facing content relevant to this PR. If no changelog is required, then add the `no-changelog-required` label.

### Affected components

- [x] Server

#### Catalog upgrade

Does this PR change Hasura Catalog version?
- [ ] No
- [ ] Yes
  - [ ] Updated docs with SQL for downgrading the catalog 

#### Metadata

Does this PR add a new Metadata feature?
- [x] No

#### GraphQL
- [x] No new GraphQL schema is generated

#### Breaking changes

- [x] There are breaking changes:

This changes the type of the `locked` column to be a `timestamp null` instead of a `boolean not null`.

### CODE REVIEW

File: server/src-lib/Hasura/Eventing/EventTrigger.hs
Comment (jberryman) [inline server/src-lib/Hasura/Eventing/EventTrigger.hs:415]:
```suggestion
                          and (l.locked IS NULL or l.locked < (NOW() - interval '30 minute'))
```

as a very non-expert on what implicit type conversions might be happening in places in postgres, this makes me more comfortable

File: server/src-lib/Hasura/RQL/DDL/EventTrigger.hs
Comment (jberryman) [inline server/src-lib/Hasura/RQL/DDL/EventTrigger.hs:161]:
```suggestion
              SELECT l.id, l.locked IS NOT NULL AND l.locked >= (NOW() - interval '30 minute')
```

File: server/src-rsr/initialise.sql
Comment (jberryman) [inline server/src-rsr/initialise.sql:305]:
```suggestion
  /* when locked IS NULL the event is unlocked and can be processed */
  locked TIMESTAMP,
```

**Comment by tirumaraiselvan** [comment]
@paf31 IIUC, we can now get rid of this? https://github.com/hasura/graphql-engine/blob/df51a8eb181cda2df5b3d5f852df6d20b046597d/server/src-lib/Hasura/App.hs#L371

File: server/src-rsr/initialise.sql
Comment (codingkarthik) [inline server/src-rsr/initialise.sql:306]: Not sure if this will affect the behaviour, but I think it will be a good idea to use `TIMESTAMPZ` instead of `TIMESTAMP` because `now()` returns a `TIMESTAMPZ` and there may be issues when comparing a `TIMESTAMPZ` value with a `TIMESTAMP` value

**Comment by paf31** [comment]
> @paf31 IIUC, we can now get rid of this?
> 
> https://github.com/hasura/graphql-engine/blob/df51a8eb181cda2df5b3d5f852df6d20b046597d/server/src-lib/Hasura/App.hs#L371

That code unlocks all events if I understand correctly. I think it might be good to keep it, so that we have a way to quickly unlock everything without having to wait 30 minutes.

**Comment by tirumaraiselvan** [comment]
> That code unlocks all events if I understand correctly. I think it might be good to keep it, so that we have a way to quickly unlock everything without having to wait 30 minutes.

We are tracking events which are in-memory and unlocking them during graceful shutdown, right? https://github.com/hasura/graphql-engine/blob/2c397f9f4f7628d6eb893974e273ecf3884d8a17/server/src-lib/Hasura/App.hs#L462 This way a proper server restart should not wait 30 mins to retry a locked event.

Shouldn't this cover most cases? Otherwise, I am not really sure what is the ROI of changing the locked column from bool to timestamp. We are managing the same amount of state as before?

**Comment by jberryman** [comment]
Mm, yeah I agree we shouldn't be doing `unlockAllEvents` on startup anymore. Formerly I guess that was the only way to ensure at-least-once delivery in the case of a hard shutdown. But now that case is taken care of by having the timeout state in durable storage.

While this is still open, do we want to think more about the timeout for (re)delivery? Maybe it should be much shorter or be relative to the http client timeout value we use? Just a detail.

**Comment by paf31** [comment]
> While this is still open, do we want to think more about the timeout for (re)delivery?

Definitely. I wasn't sure about this so I just did the most naive thing possible to start with. Any ideas?

**Comment by jberryman** [comment]
> Any ideas?

I'm not sure, and setting it too low is the real danger (sort of slow thundering herd situation... lumbering herd). Seems good for now

**Comment by jberryman** [comment]
(sorry if this was awaiting another approval by me)

**Comment by paf31** [comment]
> (sorry if this was awaiting another approval by me)

Sorry, it may have, but I'm also being a little hesitant to merge this, since it is a little difficult to test and be sure that it is the best change, for arguably little benefit. Since the original problem is (I think) now addressed, I'm thinking we can hold off until I can write some tests for this. What do you think?

**Comment by tirumaraiselvan** [comment]
I have marked this for v1.4 release since we get a beta period to test this change.

**Review by codingkarthik — APPROVED** [review]
LGTM

**Comment by jberryman** [comment]
(FWIW if you want to try using kodiak for auto-merge/update you can add the `auto-update-auto-merge` tag)

**Pull request merged**
